File: maximum_subarray.py
class Solution:
    # test cases:
    #                                i
    # [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    # maxSum = 6
    # maxSumAtCur = -2, 1, -2, 4, 3, 5, 6, 1, 5
    # maxSumAtPrev = 5
    #
    #                   i
    # [-4, -31, -30, -1]
    # maxSum = -1
    # maxSumAtCur = -4, -31, -30, -1
    # maxSumAtPrev = -1
    #
    #          i
    # [0, 0, 0]
    # maxSum = 0
    # maxSumAtCur = 0
    # maxSumAtPrev = 0
    #
    # []
    # maxSum = 0
    #
    #    i
    # [1]
    # maxSum = 1
    # maxSumAtCur = 1
    # maxSumAtPrev = 1
    #
    # optimal bottom up dp approach
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0

        # track global max subarray sum, compare against each local
        # sum of max subarray ending at each number
        maxSum, maxSumAtCur = float('-inf'), 0

        for num in nums:
            # sum of max subarray ending at current number is either
            # current number + sum of max subarray ending at previous
            # or current number itself whichever larger
            maxSumAtCur = max(maxSumAtCur + num, num)
            maxSum = max(maxSumAtCur, maxSum)

        return maxSum

File: cpp/maximum_subarray.cpp
class Solution {
public:
    // optimal bottom up dp approach
    int maxSubArray(vector<int>& nums) {
        int maxSum = INT_MIN, curMaxSum = 0;

        for (int &num : nums) {
            curMaxSum = max(curMaxSum + num, num);
            maxSum = max(curMaxSum, maxSum);
        }
        return maxSum;
    }

    // // suboptimal divide conquer approach
    // int maxSubArray(vector<int>& nums) {
    //     return maxPartitionSum(nums, 0, nums.size() - 1);
    // }

    // int maxPartitionSum(vector<int>& nums, int leftBound, int rightBound) {
    //     if (leftBound >= rightBound) {
    //         return nums[leftBound];
    //     }

    //     int middle = (leftBound + rightBound) / 2;
    //     int leftMaxSum = maxPartitionSum(nums, leftBound, middle);
    //     int rightMaxSum = maxPartitionSum(nums, middle + 1, rightBound);

    //     // find max subaray sum crossing over partition to left and
    //     // max subaray sum crossing over partition to right
    //     int curLeftSum = 0, curLeftMax = nums[middle];
    //     for (int i = middle; i >= leftBound; --i) {
    //         curLeftSum += nums[i];
    //         curLeftMax = max(curLeftSum, curLeftMax);
    //     }
    //     int curRightSum = 0, curRightMax = nums[middle + 1];
    //     for (int i = middle + 1; i <= rightBound; ++i) {
    //         curRightSum += nums[i];
    //         curRightMax = max(curRightSum, curRightMax);
    //     }

    //     // max subarray sum area in current parition is either max sum
    //     // in left or right, or max subarray sum crossing middle partition
    //     return max({ curLeftMax + curRightMax, leftMaxSum, rightMaxSum });
    // }
};
